File: 백준/Silver/15664. N과 M （10）/N과 M （10）.py
# 15664 N과 M (10) (실버 2)

import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**6)

N, M = map(int, input().strip().split())

visit = [0] * N

arr = list(map(int, input().strip().split()))
arr.sort()
st = []
def dfs(v):

    if len(st) == M:
        print(' '.join(map(str, st)))
        return
    
    log = 0
    for i in range(v, N):
        if visit[i] == 0 and log != arr[i]:
            log = arr[i]
            visit[i] = 1
            st.append(arr[i])
            dfs(i)
            visit[i] = 0
            st.pop()
    return
dfs(0)

File: 백준/Silver/15664. N과 M （10）/N과 M （10）.java
import java.util.*;
import java.io.*;

public class Main {
	static int n, m;
	static int[] arr;
	static boolean[] visited;
	
	static Stack<Integer> stack = new Stack<>();
	static StringBuilder sb = new StringBuilder();
	
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		
		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());
		arr = new int[n];
		visited = new boolean[n];
		
		st = new StringTokenizer(br.readLine());
		for (int i=0; i<n; i++) {
			arr[i] = Integer.parseInt(st.nextToken());
		}
		Arrays.sort(arr);
		
		dfs(0);
		System.out.println(sb);
	}
	
	public static void dfs(int s) {
		
		if (stack.size() == m) {
			for(int val : stack) {
				sb.append(val + " ");
			}
			sb.append("\n");
			return;
		}
		
		int log = 0;
		for (int i=s; i<n; i++) {
			if (log != arr[i] && !visited[i]) {
				log = arr[i];
				visited[i] = true;
				stack.push(arr[i]);
				dfs(i);
				stack.pop();
				visited[i] = false;
			}
			
		}
		return;
	}
}
